### PR TITLE
[SYCL][NVPTX] Create one bitcode library for NVPTX

### PR DESCRIPTION
**Pull request by MartinWehking**
Create one single bitcode library for NVPTX by compiling each libdev file into
bitcode first, linking these together and running opt on them.

Strip away metadata by reusing prepare_builtins from libclc.

Remove NVPTX bundles from the libdev object files and remove any unbundling
action spawned by the Clang driver for the SYCL toolchain when compiling for the
NVPTX backend.

Make the driver link against the single bitcode libraries for NVPTX for the SYCL
toolchain when device library linkage is not excluded.

Ensure that the clang tests check for the correctness of the new clang driver
actions and check if the driver still links the device code against the itt
device libraries when device library linkage has been excluded.

Refactor SYCLLibdevice.cmake by creating functions and grouping e.g. the same
compilation flags for a filetype together in one variable. Reuse these variables
and call functions to remove redundancies.

### CODE REVIEW

**Comment by bader** [comment]
@MartinWehking, why this change is done only for NVPTX target? Can't we make the same change for other targets as well?

**Comment by MartinWehking** [comment]
> @MartinWehking, why this change is done only for NVPTX target? Can't we make the same change for other targets as well?

Hi @bader
We also have a patch for the AMD target coming up (#15055).
Originally, there was one patch for both targets (#13930).

It is probably possible to copy the work for the SPIRV target.
However, there might be unforeseen issues.

We encountered several issues with enabling it for AMD as well.
Therefore, this might be better to be done as future work.

**Comment by steffenlarsen** [comment]
I would like @jinge90 to review this and maybe share some thoughts on the discussions above, especially pertaining to doing this for SPIR-V.

**Comment by jinge90** [comment]
> I would like @jinge90 to review this and maybe share some thoughts on the discussions above, especially pertaining to doing this for SPIR-V.

The libdevice cmake change looks good to me.
For SPV target, these is no technical block to combine several fallback SPV file into one but I doubt whether it is correct direction. The combined SPV file will lead to more jit overhead for underlying runtime(IGC, OpenCL CPU RT), this may be a performance issue for small kernels.
Thanks very much.

**Comment by jinge90** [comment]
> > I would like @jinge90 to review this and maybe share some thoughts on the discussions above, especially pertaining to doing this for SPIR-V.
> 
> The libdevice cmake change looks good to me. For SPV target, these is no technical block to combine several fallback SPV file into one but I doubt whether it is correct direction. The combined SPV file will lead to more jit overhead for underlying runtime(IGC, OpenCL CPU RT), this may be a performance issue for small kernels. 
For .obj/.o file, there is no such problem, because we use "llvm-link -only-needed" to remove all unnecessary code.
Thanks very much.

**Comment by MartinWehking** [comment]
@jinge90 

That sounds like a reasonable concern.
If there are no objections I would leave it as it is for this patch.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 1]
```diff
@@ -247,10 +254,21 @@ SYCL::getDeviceLibraries(const Compilation &C, const llvm::Triple &TargetTriple,
           C.getDriver().Diag(diag::err_drv_unsupported_option_argument)
               << A->getSpelling() << Val;
         }
-        DeviceLibLinkInfo[Val] = true && !NoDeviceLibs;
+        DeviceLibLinkInfo[Val] = !NoDeviceLibs && !IgnoreSingleLibs;
+        PrintUnusedLibWarning = IgnoreSingleLibs && !NoDeviceLibs;
       }
+      if (PrintUnusedLibWarning)
+        C.getDriver().Diag(diag::warn_ignored_clang_option)
+            << A->getSpelling() << A->getAsString(Args);
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/SYCL.cpp:262]: Can you add a test that exercises when this warning is emitted?

File: clang/test/Driver/sycl-device-lib-nvptx.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+// UNSUPPORTED: system-windows
+
+// Check if internal libraries are still linked against when linkage of all device libs is manually excluded.
+// RUN: %clangxx -ccc-print-phases -std=c++11 \
+// RUN: -fsycl -fno-sycl-device-lib=all \
+// RUN: -fsycl-targets=nvptx64-nvidia-cuda %s 2>&1 \
+// RUN: | FileCheck -check-prefix=CHK-NO-DEVLIB %s
+
+// CHK-NO-DEVLIB: [[LIB1:[0-9]+]]: input, "{{.*}}libsycl-itt-user-wrappers.bc", ir, (device-sycl, sm_50)
```
Comment (hdelan) [inline clang/test/Driver/sycl-device-lib-nvptx.cpp:13]: I would maybe also add some `CHK-NO-DEVICELIB-NOT`s to ensure the phases don't include linking with devicelib

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 1 above)
Comment (MartinWehking) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: Thanks, I also noticed that there were errors with the way these flags were treated.
There's a test case now that checks if the warning is emitted and the flags are treated correctly for NVPTX:
[clang/test/Driver/sycl-device-lib-nvptx.cpp](https://github.com/intel/llvm/pull/15048/files/e574b9a3df997472af738bf9600ac39bdc76ec11#diff-cfd66e95659b53c801d104efbf6779a187ef4049935095a7a6ef1ba3e91c49cf)

**Comment by npmiller** [comment]
ping @intel/dpcpp-cfe-reviewers 

@steffenlarsen given @jinge90 's input are you happy with this PR? It seems like it still makes sense for Nvidia and AMD but that we maybe shouldn't extend it to SPIR-V. But either way this also gives the CMake a much needed cleanup.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 3]
```diff
@@ -247,10 +267,22 @@ SYCL::getDeviceLibraries(const Compilation &C, const llvm::Triple &TargetTriple,
...
     }
   }
+
+  if (TargetTriple.isNVPTX() && IgnoreSingleLibs) {
+    LibraryList.push_back(Args.MakeArgString("devicelib--cuda.bc"));
+  }
+
+  if (IgnoreSingleLibs) {
+    return LibraryList;
+  }
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/SYCL.cpp:284]:
```suggestion
  if (TargetTriple.isNVPTX() && IgnoreSingleLibs)
    LibraryList.push_back(Args.MakeArgString("devicelib--cuda.bc"));

  if (IgnoreSingleLibs)
    return LibraryList;
```
Formatting nit.

**Review by mdtoguchi — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]

**Review by steffenlarsen — APPROVED** [review]
Let's do it! 👍

**Comment by MartinWehking** [comment]
ping @intel/llvm-gatekeepers to get this merged

**Pull request merged**